core/src/consensus/tendermint/mod.rs

```diff
@@ -279,14 +293,38 @@ impl Tendermint {
         *self.lock_change.write() = None;
         *self.proposal.write() = None;
         *self.proposal_parent.write() = Default::default();
-        *self.proposal_block.write() = None;
     }
 
     fn to_step(&self, step: Step) {
         self.extension.set_timer_step(step);
         *self.step.write() = step;
         match step {
-            Step::Propose => self.update_sealing(),
+            Step::Propose => {
+                let c = match self.client.read().as_ref().and_then(|weak| weak.upgrade()) {
+                    Some(c) => c,
+                    None => return,
+                };
+
+                let vote_step = VoteStep::new(
+                    self.height.load(AtomicOrdering::SeqCst),
+                    self.view.load(AtomicOrdering::SeqCst),
+                    Step::Propose,
+                );
+
+                if let Some(hash) = self.votes.get_block_hashes(&vote_step).first() {
+                    if let Some(header) = c.block_header(&BlockId::Hash(*hash)) {
+                        *self.proposal.write() = Some(*hash);
+                        *self.proposal_parent.write() = header.parent_hash();
+                    } else {
+                        ctrace!(ENGINE, "Proposal is received but not imported");
+                        // Proposal is received but is not verified yet.
+                        // Wait for verification.
+                        return
+                    }
+                } else {
+                    self.update_sealing()
```

Comment:
When does `get_block_hashes` return None?

Reply:
`self.votes.get_block_hashes(&vote_step).first()` is true when a node received proposal message before it entered to the "propose" step.
If a node didn't receive the proposal before enter the propose step, the value is None.

Reply:
Is it okay to call `update_sealing`?
Even if the node is not a proposer?
Or is it guaranteed that the node is a proposer when there is no received proposal?

Reply:
It has been like this before.

I agree with that we do not need to call update_sealing if the node is not a proposer.
Fixing this behavior should be another PR.